File: src/main/java/com/github/simiacryptus/aicoder/config/AppSettingsComponent.java
package com.github.simiacryptus.aicoder.config;

import com.github.simiacryptus.aicoder.openai.OpenAI_API;
import com.github.simiacryptus.aicoder.util.StyleUtil;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.ui.ComboBox;
import com.intellij.ui.components.JBCheckBox;
import com.intellij.ui.components.JBPasswordField;
import com.intellij.ui.components.JBTextField;
import com.jetbrains.rd.util.LogLevel;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import java.awt.event.ActionEvent;
import java.util.Arrays;

public class AppSettingsComponent {
    private static final Logger log = Logger.getInstance(AppSettingsComponent.class);


    @Name("Style")
    public final JBTextField style = new JBTextField();
    @SuppressWarnings("unused")
    public final JButton randomizeStyle = new JButton(new AbstractAction("Randomize Style") {
        @Override
        public void actionPerformed(ActionEvent e) {
            style.setText(StyleUtil.INSTANCE.randomStyle());
        }
    });

    @SuppressWarnings("unused")
    public final JButton testStyle = new JButton(new AbstractAction("Test Style") {
        @Override
        public void actionPerformed(ActionEvent e) {
            StyleUtil.INSTANCE.demoStyle(style.getText());
        }
    });

    @Name("Token Counter")
    public final JBTextField tokenCounter = new JBTextField();
    @SuppressWarnings("unused")
    public final JButton clearCounter = new JButton(new AbstractAction("Clear Token Counter") {
        @Override
        public void actionPerformed(ActionEvent e) {
            tokenCounter.setText("0");
        }
    });

    @SuppressWarnings("unused")
    @Name("Human Language")
    public final JBTextField humanLanguage = new JBTextField();
    @SuppressWarnings("unused")
    @Name("History Limit")
    public final JBTextField historyLimit = new JBTextField();
    @SuppressWarnings("unused")
    @Name("Developer Tools")
    public final JBCheckBox devActions = new JBCheckBox();
    @SuppressWarnings("unused")
    @Name("Suppress Progress (UNSAFE)")
    public final JBCheckBox suppressProgress = new JBCheckBox();
    @SuppressWarnings("unused")
    @Name("API Log Level")
    public final ComboBox<String> apiLogLevel = new ComboBox<>(Arrays.stream(LogLevel.values()).map(Enum::name).toArray(String[]::new));

    @SuppressWarnings("unused")
    @Name("Temperature")
    public final JBTextField temperature = new JBTextField();
    @SuppressWarnings("unused")
    @Name("Max Tokens")
    public final JBTextField maxTokens = new JBTextField();
    @SuppressWarnings("unused")
    @Name("Max Prompt (Characters)")
    public final JBTextField maxPrompt = new JBTextField();
    @SuppressWarnings("unused")
    @Name("Completion Model")
    public final JComponent model_completion = OpenAI_API.INSTANCE.getModelSelector();
    @Name("Edit Model")
    public final JComponent model_edit = OpenAI_API.INSTANCE.getModelSelector();

    @Name("API Key")
    public final JBPasswordField apiKey = new JBPasswordField();
    @SuppressWarnings("unused")
    @Name("API Base")
    public final JBTextField apiBase = new JBTextField();

    public AppSettingsComponent() {
        tokenCounter.setEditable(false);
    }

    public @NotNull JComponent getPreferredFocusedComponent() {
        return apiKey;
    }

}

File: src/main/resources/sources/kt/com/github/simiacryptus/aicoder/config/AppSettingsComponent.kt
@file:Suppress("unused")

package com.github.simiacryptus.aicoder.config

import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.fileEditor.FileEditorManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.ComboBox
import com.intellij.openapi.vfs.LocalFileSystem
import com.intellij.ui.components.JBCheckBox
import com.intellij.ui.components.JBPasswordField
import com.intellij.ui.components.JBTextField
import com.simiacryptus.jopenai.ClientUtil
import com.simiacryptus.jopenai.models.ChatModels

import org.slf4j.LoggerFactory
import java.awt.event.ActionEvent
import javax.swing.AbstractAction
import javax.swing.JButton
import javax.swing.JComponent

class AppSettingsComponent {

    @Name("Token Counter")
    val tokenCounter = JBTextField()

    @Suppress("unused")
    val clearCounter = JButton(object : AbstractAction("Clear Token Counter") {
        override fun actionPerformed(e: ActionEvent) {
            tokenCounter.text = "0"
        }
    })

    @Suppress("unused")
    @Name("Human Language")
    val humanLanguage = JBTextField()

    @Suppress("unused")
    @Name("Listening Port")
    val listeningPort = JBTextField()

    @Suppress("unused")
    @Name("Listening Endpoint")
    val listeningEndpoint = JBTextField()

    @Suppress("unused")
    @Name("Suppress Errors")
    val suppressErrors = JBCheckBox()

    @Suppress("unused")
    @Name("Model")
    val modelName = ComboBox<String>()

    @Suppress("unused")
    @Name("Enable API Log")
    val apiLog = JBCheckBox()

    @Suppress("unused")
    val openApiLog = JButton(object : AbstractAction("Open API Log") {
        override fun actionPerformed(e: ActionEvent) {
            ClientUtil.auxiliaryLog?.let {
                val project = ApplicationManager.getApplication().runReadAction<Project> {
                    com.intellij.openapi.project.ProjectManager.getInstance().openProjects.firstOrNull()
                }
                if (it.exists()) {
                    ApplicationManager.getApplication().invokeLater {
                        val virtualFile = LocalFileSystem.getInstance().refreshAndFindFileByIoFile(it)
                        FileEditorManager.getInstance(project!!).openFile(virtualFile!!, true)
                    }
                } else {
                    log.warn("Log file not found: ${it.absolutePath}")
                }
            }
        }
    })


    @Suppress("unused")
    @Name("Developer Tools")
    val devActions = JBCheckBox()

    @Suppress("unused")
    @Name("Edit API Requests")
    val editRequests = JBCheckBox()

    @Suppress("unused")
    @Name("Temperature")
    val temperature = JBTextField()

    @Name("API Key")
    val apiKey = JBPasswordField()

    @Suppress("unused")
    @Name("API Base")
    val apiBase = JBTextField()

    @Name("File Actions")
    var fileActions = ActionTable(AppSettingsState.instance.fileActions.actionSettings.values.map { it.copy() }
        .toTypedArray().toMutableList())

    @Name("Editor Actions")
    var editorActions = ActionTable(AppSettingsState.instance.editorActions.actionSettings.values.map { it.copy() }
        .toTypedArray().toMutableList())

    init {
        tokenCounter.isEditable = false
        this.modelName.addItem(ChatModels.GPT35Turbo.modelName)
        this.modelName.addItem(ChatModels.GPT4.modelName)
        this.modelName.addItem(ChatModels.GPT4Turbo.modelName)
    }

    val preferredFocusedComponent: JComponent
        get() = apiKey

    class ActionChangedListener {
        fun actionChanged() {
        }
    }

    companion object {
        private val log = LoggerFactory.getLogger(AppSettingsComponent::class.java)
        //val ACTIONS_TOPIC = Topic.create("Actions", ActionChangedListener::class.java)
    }
}
